File: python3/find_mode_in_binary_search_tree.py
from .model import TreeNode

"""
Space   : O(n)
Time    : O(n)
"""

class Solution:
    def findMode(self, root: TreeNode) -> List[int]:
        if not root:
            return[]

        mem = []

        stack = [root]

        while stack:
            node = stack.pop()
            mem.append(node.val)
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)

        hashmap = {}
        max_count = 0

        for i in mem:
            if i not in hashmap:
                hashmap[i] = 1
            else:
                hashmap[i] += 1
            max_count = max(max_count, hashmap[i])

        mode = []
        for k, v in hashmap.items():
            if v == max_count:
                mode.append(k)

        return mode

File: golang/find_mode_in_binary_search_tree.go
package main

import "golang/model"

/*
Space	: O(n)
Time	: O(n)
*/

func findMode(root *model.TreeNode) []int {
	hashmap := map[int]int{}
	ans := []int{}

	stack := []*model.TreeNode{root}
	for len(stack) > 0 {
		var temp []*model.TreeNode

		for _, node := range stack {
			hashmap[node.Val]++
			if node.Left != nil {
				temp = append(temp, node.Left)
			}
			if node.Right != nil {
				temp = append(temp, node.Right)
			}
		}

		stack = temp
	}

	max_count := 0
	for _, v := range hashmap {
		max_count = Max(max_count, v)
	}

	for k, v := range hashmap {
		if v == max_count {
			ans = append(ans, k)
		}
	}

	return ans
}
